bitcoin/script/miniscript.h

```diff
@@ -1014,224 +1014,384 @@ bool DecomposeScript(const CScript& script, std::vector<std::pair<opcodetype, st
-        in += 7;
-        return MakeNodeRef<Key>(NodeType::HASH160, in[-6].second);
-    }
-    if (last - in >= 2 && in[0].first == OP_CHECKSIG) {
-        ++in;
-        auto sub = DecodeSingle<Key>(in, last, ctx);
-        if (!sub) return {};
-        return MakeNodeRef<Key>(NodeType::WRAP_C, Vector(std::move(sub)));
-    }
-    if (last - in >= 3 && in[0].first == OP_BOOLAND) {
-        ++in;
-        auto sub1 = DecodeWrapped<Key>(in, last, ctx);
-        if (!sub1) return {};
-        auto sub2 = DecodeSingle<Key>(in, last, ctx);
-        if (!sub2) return {};
-        return MakeNodeRef<Key>(NodeType::AND_B, Vector(std::move(sub2), std::move(sub1)));
-    }
-    if (last - in >= 3 && in[0].first == OP_BOOLOR) {
-        ++in;
-        auto sub1 = DecodeWrapped<Key>(in, last, ctx);
-        if (!sub1) return {};
-        auto sub2 = DecodeSingle<Key>(in, last, ctx);
-        if (!sub2) return {};
-        return MakeNodeRef<Key>(NodeType::OR_B, Vector(std::move(sub2), std::move(sub1)));
-    }
-    if (last - in >= 2 && in[0].first == OP_VERIFY) {
-        ++in;
-        auto sub = DecodeSingle<Key>(in, last, ctx);
-        if (!sub) return {};
-        return MakeNodeRef<Key>(NodeType::WRAP_V, Vector(std::move(sub)));
-    }
-    if (last - in >= 2 && in[0].first == OP_0NOTEQUAL) {
-        ++in;
-        auto sub = DecodeSingle<Key>(in, last, ctx);
-        if (!sub) return {};
-        return MakeNodeRef<Key>(NodeType::WRAP_N, Vector(std::move(sub)));
-    }
-    if (last > in && in[0].first == OP_ENDIF) {
-        ++in;
-        if (last - in == 0) return {};
-        NodeRef<Key> sub1;
-        sub1 = DecodeMulti<Key>(in, last, ctx);
-        if (!sub1) return {};
-        bool have_else = false;
-        NodeRef<Key> sub2;
-        if (last - in == 0) return {};
-        if (in[0].first == OP_ELSE) {
-            ++in;
-            have_else = true;
-            sub2 = DecodeMulti<Key>(in, last, ctx);
-            if (!sub2) return {};
-        }
-        if (last - in == 0 || (in[0].first != OP_IF && in[0].first != OP_NOTIF)) return {};
-        bool negated = (in[0].first == OP_NOTIF);
-        ++in;
-
-        if (!have_else && !negated) {
-            if (last > in && in[0].first == OP_DUP) {
+inline NodeRef<Key> DecodeScript(I& in, I last, const Ctx& ctx) {
+    // The two integers are used to hold state for thresh()
+    std::vector<std::tuple<DecodeContext, int64_t, int64_t>> to_parse;
+    std::vector<NodeRef<Key>> constructed;
+
+    // There may be implicit and_v expressions [X] [Y] or s: wrapper SWAP [X]
+    to_parse.emplace_back(DecodeContext::MaybeAndV, -1, -1);
+    to_parse.emplace_back(DecodeContext::MaybeSwap, -1, -1);
+    to_parse.emplace_back(DecodeContext::Expression, -1, -1);
+
+    while (!to_parse.empty()) {
+
+        // Get the current context we are decoding within
+        DecodeContext cur_context = std::get<0>(to_parse.back());
```

Comment:
With C++17, this and the next two lines can be written more concisely as `auto [cur_context, n, k] = to_parse.back();`.

Reply:
Nice, I wasn't sure if you wanted to require c++17 support because Makefile has `-std=c++11`